[APM] Link Transaction to Infra & Logging with action menu

ogupte opened the pull request:
Fixes #26574 by adding an action menu to Transaction details and flyout populated with applicable links to the Infra & Logging UIs

<img width="1030" alt="screen shot 2018-12-17 at 1 33 30 am" src="https://user-images.githubusercontent.com/1967266/50078470-eda3bb80-019b-11e9-8945-ecd6936058b9.png">

ogupte committed:
[APM] fixes #26574 by adding an Action menu which links a transaction to the Infra & Logging UIs

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4059/)

jasonrhodes commented:
FWIW I think @elastic/eui-design just "fixed" the context menu header by making it never have a grey background (so the triangle color is never weird like in the screenshot), but I'm not sure if that's out yet. You may try rebasing from master and doing a kbn bootstrap.

snide commented:
> but I'm not sure if that's out yet

It will once this is in https://github.com/elastic/kibana/pull/27009

sqren commented on x-pack/plugins/apm/public/components/shared/TransactionActionMenu/TransactionActionMenu.tsx:
```diff
@@ -0,0 +1,204 @@
+
+export class TransactionActionMenu extends React.Component<
+  TransactionActionMenuProps,
+  TransactionActionMenuState
```
Can we just name these `Props` and `State` or are there other similar interfaces causing naming conflicts/confusion?

sqren commented on x-pack/plugins/apm/public/components/shared/TransactionActionMenu/TransactionActionMenu.tsx:
```diff
@@ -0,0 +1,204 @@
+    // @ts-ignore - TODO remove this ignore when support for TransactionV1 is dropped
+    const podId = oc(transaction).kubernetes.pod.uid();
+    // @ts-ignore - TODO remove this ignore when support for TransactionV1 is dropped
+    const containerId = oc(transaction).container.id();
```
As discussed: `oc` doesn't work on the client :'(

sqren commented on x-pack/plugins/apm/public/utils/url.tsx:
```diff
@@ -142,6 +142,27 @@ export function RelativeLinkComponent({
   );
 }
 
+export function getUnconnectedKibanaHref(
```
`getKibanaHref`

sqren commented on x-pack/plugins/apm/public/components/shared/TransactionActionMenu/TransactionActionMenu.tsx:
```diff
@@ -0,0 +1,204 @@
+        icon="discoverApp"
+        href={discoverTransactionHref}
+        key="discover-transaction"
+      >
```
You also need to specify the index pattern like in https://github.com/elastic/kibana/blob/26c77eb25edd5129265208c24dd3a6ab68a2885b/x-pack/plugins/apm/public/components/shared/DiscoverButtons/DiscoverButton.tsx#L34-L36

Instead of duplicating that, perhaps you can make some share method for doing this.

Note: `getAPMIndexPattern` calls a memoized function, so there is no need to avoid additional calls.

ogupte replied:
good catch. I abstracted this logic into an async function `setApmIndexPatternQuery(query)` and use it in both places

ogupte committed:
[APM]
- remove link to host metrics filterd by trace id until supported
- replace ts-optchain with _.get
- remove old, unreferenced ActionMenu dead code
- add unit tests

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4283/)

ogupte committed:
[APM] Make sure the apm index pattern filter is passed thru via the Discover link

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4297/)

sqren commented on x-pack/plugins/apm/public/components/shared/DiscoverButtons/DiscoverButton.tsx:
```diff
@@ -6,34 +6,44 @@
+
+  query._a.index = id;
+
+  return query;
```
Mutating `query` might give some surprises down the road. Can you make it immutable, and rename to `getQueryWithApmIndexPattern `?

sqren commented on x-pack/plugins/apm/public/components/shared/TransactionActionMenu/TransactionActionMenu.tsx:
```diff
@@ -0,0 +1,200 @@
+    const discoverQuery = await setApmIndexPatternQuery(
+      this.state.discoverQuery
+    );
+    this.setState({ discoverQuery });
```
How do you handle if `this.props.transaction` changes?

ogupte replied:
my first fix to deal with this was to add more lifecycle methods, then it become more complicated, so then i made getQueryWithApmIndexPattern synchronous, and only do the query object building, and had to rely on each component to do the async/state management stuff, which kind of defeats the purpose of the abstraction.

So in the end i created a new component `QueryWithIndexPattern` which handles all the async, state management, and query object transform. I ended up implementing it as a FaCC so that it the final query object can be used in either case (passed as a prop in DiscoverButton, or used to build the href in TransactionActionMenu)

LMK what you think.

ogupte committed:
[APM] refactored DiscoverButton and TransactionActionMenu to use new QueryWithIndexPattern component to pass along the correct index pattern via the discover link

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4447/)

sqren commented on x-pack/plugins/apm/public/components/shared/TransactionActionMenu/TransactionActionMenu.tsx:
```diff
@@ -0,0 +1,194 @@
+  public render() {
+    const { transaction, location } = this.props;
+    return (
+      <QueryWithIndexPattern query={getDiscoverQuery(transaction)}>
```
It's pretty crazy that doing async rendering has always been this hard. Hooks cannot come soon enough...

sqren replied:
Btw. good job. This is definitely a cleaner solution!

sqren commented on x-pack/plugins/apm/public/utils/url.tsx:
```diff
@@ -150,6 +150,25 @@ export function RelativeLinkComponent({
+    hash: `${hash}?${search}`
+  });
+  return href;
+}
```
This is great. I think we are gonna be happy about this decoupling.
